### PR TITLE
fix(developer): kmlmc generates invalid json with smp

### LINKED ISSUE #2374: [Developer] kmlmc generates invalid code for SMP characters
The ES5 engine doesn't like duplicate property names on objects, and there's a minor bug generating them within the lexical model compiler.

In particular, two SMP characters have the same leading surrogate:
- 🇸
- 🇺

It appears that when compiling SMP strings for Trie models, a node is being generated for the leading surrogate, followed by a node for the completed surrogate pair.  Unfortunately, a unique copy of the leading surrogate is generated for each surrogate pair, allowing duplicate object properties to exist if two SMP characters (like the two above) have the same leading surrogate.

_Originally posted by @jahorton in https://github.com/keymanapp/keyman/pull/2358#issuecomment-555909909_

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #2374.

Updated:

This PR updates node to 12.0, which [adds support](https://v8.dev/features/well-formed-json-stringify) for the [well-formed JSON.stringify proposal](https://github.com/tc39/proposal-well-formed-stringify)

The issue arises because `JSON.stringify` was replacing unpaired surrogates with `U+FFFD` (Unicode replacement character).

The solution is to keep the unpaired surrogates but store them in `\uxxxx` notation in the .js to avoid unpaired surrogates as code units. This is a toolchain update + a unit test.

### CODE REVIEW

**Comment by eddieantonio** [comment]
The Trie implementation stores UTF-16 code units on each node --- intentionally splitting apart high and low surrogates. This is for easy of indexing (no need for conditional logic at the cost of extra space for astral code points).

So the fix would be to use escapes for unmatched surrogates, I suppose?

**Comment by mcdurdin** [comment]
The main problem is that lead surrogates seem to be repeated in the array.
I’m not yet clear why this is happening because as far as I can tell that
shouldn’t happen.

You can try building the example.qaa.smp model to see the result and
observe the repetition.

**Comment by mcdurdin** [comment]
I think it's because streaming out is replacing the unpaired surrogates with `U+FFFD` (Unicode replacement character). It's remarkably hard to be confident of the content of the file but now I've loaded it with a hex editor I can see `EF BF BD` which is UTF-8 for `FFFD`. 🤣

So yes, replacing that output with `\uBxxx` will do the job. Now, to figure out how -- may be a little tricky.

**Comment by mcdurdin** [comment]
See also https://github.com/tc39/proposal-well-formed-stringify and https://v8.dev/features/well-formed-json-stringify

**Comment by mcdurdin** [comment]
Works on my machine. Need to make sure the build chain is right now. node versions can be a bit tricky!

File: developer/js/package.json [hunk 1]
```diff
@@ -36,7 +36,7 @@
   },
   "dependencies": {
     "commander": "^3.0.0",
-    "node": "^11.7.0",
+    "node": "^12.0.0",
```
Comment (eddieantonio) [inline developer/js/package.json]: I believe this change should be in the `engines` key in `package.json`, not in dependencies See: https://docs.npmjs.com/files/package.json#engines

Comment (mcdurdin) [reply developer/js/package.json]: Okay, I've added the `engines` key but also left this one the same. I don't think that's a problem, right?

Comment (eddieantonio) [reply developer/js/package.json]: It's not a problem, _per se_, but the `node` npm package is [an actual real package](https://www.npmjs.com/package/node) and all it does is it downloads a binary of node for the current platform. It doesn't declare a dependency on node itself; it _installs_ a version of node! Try ls'ing `node_modules/node/bin`.

**Comment by mcdurdin** [comment]
> It's not a problem, per se, but the node npm package is an actual real package and all it does is it downloads a binary of node for the current platform. It doesn't declare a dependency on node itself; it installs a version of node! Try ls'ing node_modules/node/bin.

Well ... actually this solves a weird issue on my machine. Running build.js in the developer/js folder ended up with a wild [This: command not found](https://github.com/nodejs/node/issues/29287) error (not identical in behaviour to that link, but close enough), because under bash on Windows, `node_modules/node/bin/node` would be executed in preference to `node_modules/node/bin/node.exe`. `node` was a file containing the text `This file has been intentionally left blank` (or something to that effect). Removing node as a dependency resolves this crazyness.

**Comment by eddieantonio** [comment]
Awesome! LGTM! :shipit:

**Pull request merged**